module/extension/Director/src/component/DirectorTask.hpp

```diff
@@ -0,0 +1,109 @@
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2023 NUbots <nubots@nubots.net>
+ */
+
+#ifndef MODULE_EXTENSION_DIRECTOR_DIRECTORTASK_HPP
+#define MODULE_EXTENSION_DIRECTOR_DIRECTORTASK_HPP
+
+#include <map>
+#include <typeindex>
+#include <vector>
+
+#include "Director.hpp"
+
+#include "extension/Behaviour.hpp"
+
+namespace module::extension::component {
+
+    /**
+     * A Task to use in the Director algorithm based on the BehaviourTask, with added state fields for the Director
+     * algorithm
+     */
+    struct DirectorTask {
+
+        /**
+         * Construct a new Task object for use in the director algorithm using a given BehaviourTask
+         *
+         * @param task                  the BehaviourTask to use as a base
+         * @param optional_             whether this task is optional or not
+         */
+        DirectorTask(::extension::behaviour::commands::BehaviourTask task, bool dying_) : dying(dying_) {
+            type              = task.type;
+            root_type         = task.root_type;
+            requester_id      = task.requester_id;
+            requester_task_id = task.requester_task_id;
+            data              = task.data;
+            name              = task.name;
+            priority          = task.priority;
+            optional          = task.optional;
+        }
+
+        /**
+         * Construct a new Task object for use in the director algorithm
+         *
+         * @param type_                 the type that this task is for
+         * @param root_type_            a secondary type to use if this is a root task
+         * @param requester_id_         the id of the reaction that emitted this task
+         * @param requester_task_id_    the task_id of the reaction task that emitted this
+         * @param data_                 the task data to be sent to the provider
+         * @param name_                 a string name for this task for use in debugging
+         * @param priority_             the priority that this task is to run with
+         * @param optional_             whether this task is optional or not
+         * @param dying_                whether this task is currently about to be removed
+         */
+        DirectorTask(std::type_index type_,
+                     std::type_index root_type_,
```

Comment:
I think that at the point where you make this DirectorTask you no longer need the root type...